src/target/hc32l110.c

```diff
@@ -0,0 +1,236 @@
+		break;
+	case 32768:
+		target_add_ram(target, 0x2000000, 4096);
+		break;
+	default:
+		return false;
+	}
+
+	target->driver = "HC32L110";
+	target->enter_flash_mode = hc32l110_enter_flash_mode;
+	target->mass_erase = hc32l110_mass_erase;
+
+	hc32l110_add_flash(target, flash_size);
+	return true;
+}
+
+/* Executes the magic sequence to unlock the CR register */
+static void hc32l110_flash_cr_unlock(target_s *const target)
+{
+	target_mem_write32(target, HC32L110_FLASH_BYPASS, 0x5a5aU);
+	target_mem_write32(target, HC32L110_FLASH_BYPASS, 0xa5a5U);
+}
+
+static bool hc32l110_check_flash_completion(target_s *const target, const uint32_t timeout_ms)
+{
+	platform_timeout_s timeout;
+	platform_timeout_set(&timeout, timeout_ms);
+	uint32_t status = HC32L110_FLASH_CR_BUSY;
+	while (status & HC32L110_FLASH_CR_BUSY) {
+		status = target_mem_read32(target, HC32L110_FLASH_CR);
+		if (target_check_error(target) || platform_timeout_is_expired(&timeout))
+			return false;
+	}
+	return true;
+}
+
+/* Lock the whole flash */
+static void hc32l110_slock_lock_all(target_s *const target)
+{
+	hc32l110_flash_cr_unlock(target);
+	target_mem_write32(target, HC32L110_FLASH_SLOCK, 0);
+}
+
+/* Unlock the whole flash for writing */
+static void hc32l110_slock_unlock_all(target_s *const target)
+{
+	hc32l110_flash_cr_unlock(target);
+	target_mem_write32(target, HC32L110_FLASH_SLOCK, 0xffffU);
+}
+
+static bool hc32l110_enter_flash_mode(target_s *const target)
+{
+	target_reset(target);
+
+	/*
+	 * The Flash controller requires the core's program counter to be
+	 * outside of the Flash to unlock all regions of the Flash
+	 * (Whatever sector it is left in becomes stuck in a locked state)
+	 */
+	const uint32_t pc = 0xfffffffeU;
+	return target_reg_write(target, REG_PC, &pc, sizeof(pc)) == sizeof(pc);
+}
+
+static bool hc32l110_flash_prepare(target_flash_s *const flash, const flash_operation_e op)
+{
+	hc32l110_flash_cr_unlock(flash->t);
+
+	switch (op) {
+	case FLASH_OPERATION_WRITE:
+		target_mem_write32(flash->t, HC32L110_FLASH_CR, HC32L110_FLASH_CR_OP_PROGRAM);
+		break;
+	case FLASH_OPERATION_ERASE:
+		target_mem_write32(flash->t, HC32L110_FLASH_CR, HC32L110_FLASH_CR_OP_ERASE_SECTOR);
+		break;
+	default:
+		DEBUG_WARN("unsupported operation %d", op);
+		return false;
+	}
+
+	hc32l110_slock_unlock_all(flash->t);
+	return true;
+}
+
+static bool hc32l110_flash_done(target_flash_s *const flash, const flash_operation_e op)
+{
+	(void)op;
+	hc32l110_slock_lock_all(flash->t);
+	return true;
+}
+
+static bool hc32l110_flash_erase(target_flash_s *const flash, const target_addr_t addr, const size_t length)
+{
+	(void)length;
+	/* The Flash controller automatically erases the whole sector after one write operation */
+	target_mem_write32(flash->t, addr, 0);
+
+	const bool result = hc32l110_check_flash_completion(flash->t, 1000);
+	if (!result)
+		DEBUG_ERROR("Failed to erase %" PRIx32 " bytes at 0x%08" PRIx32 "\n", (uint32_t)length, addr);
```

Comment:
Isn't this debug message redundant? `target_flash_erase` does a `DEBUG_ERROR("Erase failed at %" PRIx32 "\n", local_start_addr);` on erase failure

Reply:
You're right, we weren't thinking about the target layer, only what was in front of us from the original contributor.